lib/recursive-methods.rb

```diff
@@ -1,49 +1,115 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: n
+# Space complexity: n
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+    return 1 if n == 1
+    return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: n^2
+# Space complexity: n^2
+def reverse(s) #was the answer to this supposed to be different from the one below??
+    if s.length == 0
+        return s
+    else
+        return reverse(s[1..-1]) +s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: n
+# Space complexity: n
 def reverse_inplace(s)
```

Comment:
👍 This is the same as the previous method and it's not in place.

Consider passing the front and rear indexes and swapping elements at each index.

```ruby
def reverse_inplace(s, front = 0, last = s.length - 1)
  return s if front >= last

  temp = s[last]
  s[last] = s[front]
  s[front] = temp

  return reverse_inplace(s, front + 1, last - 1)
end
```